app/controllers/searches_controller.rb

```diff
@@ -11,10 +11,56 @@ def index
       end
 
       render :index
-    # else
-    #   flash.now[:alert] = 'Please enter a city name and select a type to search'
-    #   @places = []
-    #   # render :index
+      # else
+      #   flash.now[:alert] = 'Please enter a city name and select a type to search'
+      #   @places = []
+      #   # render :index
     end
   end
-end
\ No newline at end of file
+
+  def add_to_favorites
+    place_id = params[:place][:id].to_i
+    place = params[:place][:place_type].singularize
+
+    case place
+    when "Attraction"
+      if current_user.attraction_ids.include?(place_id)
+        flash[:alert] = failure_message(place)
+      else
+        current_user.attractions << Attraction.find(place_id)
+        flash[:notice] = success_message(place)
+      end
+    when "Hotel"
+      if current_user.hotel_ids.include?(place_id)
+        flash[:alert] = failure_message(place)
+      else
+        current_user.hotels << Hotel.find(place_id)
+        flash[:notice] = success_message(place_id)
+      end
+    when "Restaurant"
+      if current_user.restaurant_ids.include?(place_id)
+        flash[:alert] = failure_message(place)
+      else
+        current_user.hotels << Restaurant.find(place_id)
+        flash[:notice] = success_message(place)
+      end
+    else
+      raise PlaceTypeInvalid.new("#{place} is not supported at the moment")
```

Comment:
When you raise an error, there should be a corresponding `rescue_from` or `rescue` so that the user sees a proper error message instead of a generic 500 error page. The exception message doesn't get to the user by default.